sys/vm/swap_pager.c

```diff
@@ -2167,18 +2168,34 @@ swp_pager_meta_cheri_get_tags(vm_page_t page)
 	    i < (swidx + 1) * BITS_PER_TAGS_PER_PAGE; i++) {
 		p = scan;
 		for (t = sb->swb_tags[i]; t != 0; t >>= j) {
+			mark_capdirty = 1;
 			j = ffsl((long)t);
 			cheri_restore_tag(p + j - 1);
 			p += j;
 		}
 		scan += 8 * sizeof(uint64_t);
 	}
+
+	/*
+	 * Because we stored through the direct region we may have bypassed
+	 * the MMU and all its implicit capdirty tracking.  If we stored a
+	 * capability, mark the page as dirty.
+	 *
+	 * XXX-NWF Adjust plumbing to verify that this object is, in fact,
+	 * permitted to store tags!
+	 */
+	if (mark_capdirty) {
+		page->oflags |= VPO_CAPSTORE;
```

Comment:
This looks dubious; just because there were tags in the page when it was swapped out doesn't mean the application should be able to store more. I believe you want to be checking OBJ_NOSTORETAGS isn't set in `page->object->flags`?.